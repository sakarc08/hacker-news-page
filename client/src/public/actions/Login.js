import { USER_LOGGEDIN, NOT_AUTHENTICATED, USER_LOGOUT, SET_ALERT, CLEAR_ALERT } from './types'
import axios from 'axios';

export const loginUser = ({ email, password }) => async dispatch => {
    try {
        const config = {
            headers: {
                'content-type': "application/json"
            }
        }

        const body = JSON.stringify({ email, password });

        const result = await axios.post('/api/user/login', body, config);
        dispatch({
            type: USER_LOGGEDIN,
            payload: result.data
        });

    } catch (error) {
        console.log(error.mesage);
        error.response.data.errors.forEach(item => {
            dispatch({
                type: SET_ALERT,
                payload: { message: item.msg, type: 'danger'}
            })

            setTimeout(() => {
                dispatch({
                    type: CLEAR_ALERT
                })
            }, 5000)
        })
    }
}

export const loadUser = () => async dispatch => {
    try {
        const result = await axios.get('/api/user');
        dispatch({ type: USER_LOGGEDIN, payload: result.data });
    } catch (error) {
        console.log(error.message)
        dispatch({
            type: NOT_AUTHENTICATED
        })
    }
}

export const logoutUser = () => async dispatch => {
    try {
        dispatch({
            type: USER_LOGOUT,
            payload: false
        });

        dispatch({
            type: SET_ALERT,
            payload: { message: 'You logged out successfully ', type: 'success'}
        })

        setTimeout(() => {
            dispatch({
                type: CLEAR_ALERT
            })
        }, 5000)

    } catch (error) {
        console.log(error.mesage);
        dispatch({
            type: SET_ALERT,
            payload: { message: error.response.data.message, type: 'danger'}
        })
        
    }
}