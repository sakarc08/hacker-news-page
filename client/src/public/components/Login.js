import React, { useState, Fragment } from 'react'
import PropTypes from 'prop-types'
import { loginUser } from '../actions/Login'
import { connect } from 'react-redux';
import StoryBoardRoute from './StoryBoardRoute'
import { Link, Redirect } from 'react-router-dom';
import TextField from '@material-ui/core/TextField'
import Button from '@material-ui/core/Button'
import Label from '@material-ui/core/InputLabel'
import styled from 'styled-components'

const Form = styled.form`
        width: 50%;
        margin: 10% auto;
    `

const FormField = styled(TextField)`
    margin-top: 30px;
`

const Login = ({ user, isLoggedIn, loginUser }) => {

    const [formData, setformData] = useState({
        password: '',
        email: '',
    });

    const { password, email } = formData;
    const onChange = (e) => {
        setformData({ ...formData, [e.target.name]: e.target.value});
    }

    const onSubmit = (e) => {
        e.preventDefault();
        loginUser({ password, email })
    }

    { return isLoggedIn ? <Redirect to='/storyboard' /> : (<Fragment>
            <Form onSubmit={onSubmit}>
                <FormField
                    value={email}
                    name="email"
                    variant="outlined"
                    required
                    fullWidth
                    onChange={onChange}
                    id="email"
                    label="Email"
                    
                />
              <FormField
                    value={password}
                    name="password"
                    variant="outlined"
                    required
                    fullWidth
                    onChange={onChange}
                    id="password"
                    label="Password"
                    type="password"
              />

                <Button
                    type="submit"
                    variant="contained"
                    color="primary"
                >
                    Login
                </Button>

                Dont have account yet ?? <Link to="/signUp" > Create </Link> one now
            </Form>
        </Fragment>)
    }
}

Login.propTypes = {
    loginUser: PropTypes.func.isRequired,
}

const mapStateToProps = (state) => {
    // console.log(state)
    const { user, isLoggedIn } = state.userDetails
    return {
        user,
        isLoggedIn
    }
}

export default connect(mapStateToProps, { loginUser })(Login)
