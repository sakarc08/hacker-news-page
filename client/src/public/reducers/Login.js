import { USER_LOGGEDIN,NOT_AUTHENTICATED, USER_LOGOUT } from "../actions/types";

const initialState = {
    isLoggedIn: false,
    loading: true,
    user: null
}

export const LoginReducer = (state=initialState, action) => {
    const { payload, type } = action;

    switch (type) {
        case NOT_AUTHENTICATED:
            return { ...state, loading: false }
        case USER_LOGGEDIN:
            if(payload.token) localStorage.setItem('token', payload.token)
            return { ...state, loading: false, user: payload.user, isLoggedIn: true }
        case USER_LOGOUT: 
            localStorage.removeItem('token')
            return { ...state, loading: false, user: null, isLoggedIn: payload }
        default:
            return state
    }
}