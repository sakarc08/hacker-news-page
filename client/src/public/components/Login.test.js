import React from 'react';
import ReactDOM from 'react-dom';
import { Router } from 'react-router-dom';
import {Login} from './Login';
import renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { createMemoryHistory } from 'history';

describe('Login Component render', () => {
    const props = {
        isLoggedIn: false,
        loginUser: () => {},
        alerts: []
    }
    const history = createMemoryHistory();
    it('renders Login correctly', () => {
        const tree = renderer.create(<Router history={history}><Login {...props}></Login></Router>).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('returns null if user is loggedin', () => {
        props.isLoggedIn = true;
        const tree = renderer.create(<Router history={history}><Login {...props}></Login></Router>).toJSON();
        expect(tree).toBe(null);
    })
})