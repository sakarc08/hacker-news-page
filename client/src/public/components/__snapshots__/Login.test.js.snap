// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component render renders Login correctly 1`] = `
Array [
  <h1
    className="sc-AxiKw jwcsqL"
  >
    Welcome to HackerNews
  </h1>,
  <form
    className="sc-AxjAm lhjEdx"
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root sc-AxirZ jCdlai MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink={false}
        htmlFor="email"
        id="email-label"
      >
        Email
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="email"
          name="email"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-61"
          >
            <span>
              Email
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root sc-AxirZ jCdlai MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink={false}
        htmlFor="password"
        id="password-label"
      >
        Password
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="password"
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="password"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-61"
          >
            <span>
              Password
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="sc-AxhCb iHjNrQ"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-AxhUy bmpwMx MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Login
        </span>
      </button>
      <a
        className="sc-AxgMl kMXZGD"
        href="/signup"
        onClick={[Function]}
      >
         Create Account 
      </a>
    </div>
  </form>,
]
`;
